File: java/core/src/main/java/com/github/fernthedev/lightchat/core/packets/Packet.java
package com.github.fernthedev.lightchat.core.packets;

import com.github.fernthedev.lightchat.core.encryption.codecs.AcceptablePacketTypes;
import lombok.NonNull;
import lombok.ToString;

import java.lang.reflect.Modifier;

@ToString
public abstract class Packet implements AcceptablePacketTypes {
    private static final long serialVersionUID = -5039841570298012421L;


    public String getPacketName() {
        if (packetName == null) packetName = getClass().getAnnotation(PacketInfo.class).name();
        return packetName;
    }

    protected transient @NonNull String packetName;

    public static String getPacketName(Class<? extends Packet> packet) {
        if (packet.equals(Packet.class)) throw new IllegalArgumentException("The class cannot be " + Packet.class.getName());

        if (!Packet.class.isAssignableFrom(packet)) throw new IllegalStateException("Packet " + packet.getName() + " must extend " + Packet.class.getName());

        if (Modifier.isAbstract(packet.getModifiers()) || packet.isInterface()) throw new IllegalArgumentException("The class cannot be abstract or interface.");

        if (!packet.isAnnotationPresent(PacketInfo.class)) throw new IllegalStateException("Packet " + packet.getName() + " must have a packet info annotation");

        return packet.getAnnotation(PacketInfo.class).name();
    }


    protected Packet() {
        if (!getClass().isAnnotationPresent(PacketInfo.class)) throw new IllegalStateException("Packet must have a packet info annotation");

        packetName = getClass().getAnnotation(PacketInfo.class)

File: java/core/src/main/java/com/github/fernthedev/lightchat/core/packets/Packet.kt
package com.github.fernthedev.lightchat.core.packets

import com.github.fernthedev.lightchat.core.codecs.AcceptablePacketTypes
import java.lang.reflect.Modifier
import java.util.concurrent.ConcurrentHashMap

abstract class Packet protected constructor() : AcceptablePacketTypes {


    private val packetInfo: PacketInfo
        get() {
            check(javaClass.isAnnotationPresent(PacketInfo::class.java)) { "Class ${javaClass.name} does not have PacketInfo" }
            return javaClass.getAnnotation(PacketInfo::class.java)
        }

    val packetName: String
        get() = packetInfo.name

    init {
        check(javaClass.isAnnotationPresent(PacketInfo::class.java)) { "Packet must have a packet info annotation" }
    }

    override fun toString(): String {
        return "Packet(packetName=" + this.packetName + ")"
    }

    companion object {
        private const val serialVersionUID = -5039841570298012421L
        private val names: MutableMap<Class<Packet>, String> = ConcurrentHashMap()

        fun getPacketName(packet: Class<out Packet>): String {
            require(packet != Packet::class.java) { "The class cannot be " + Packet::class.java.name }
            check(Packet::class.java.isAssignableFrom(packet)) { "Packet " + packet.name + " must extend " + Packet::class.java.name }

            require(!(Modifier.isAbstract(packet.modifiers) || packet.isInterface)) { "The class cannot be abstract or interface." }
            check(packet.isAnnotationPresent(PacketInfo::class.java)) { "Packet " + packet.name + " must have a packet info annotation" }

            return packet.getAnnotation(PacketInfo::class.java).name
        }
    }
}